param/__init__.py

```diff
@@ -1249,31 +1282,24 @@ def __init__(self, objects=None, default=None, instantiate=False,
             elif isinstance(objects, list):
                 autodefault = objects[0]
 
-        default = autodefault if (not empty_default and default is None) else default
+        default = autodefault if (not empty_default and default is Undefined) else default
 
-        if objects is None:
-            objects = []
         if isinstance(objects, collections_abc.Mapping):
             self.names = objects
             self.objects = list(objects.values())
         else:
             self.names = None
             self.objects = objects
         self.compute_default_fn = compute_default_fn
-
-        if check_on_set is not None:
-            self.check_on_set = check_on_set
-        elif len(objects) == 0:
-            self.check_on_set = False
-        else:
-            self.check_on_set = True
+        self.check_on_set = check_on_set
 
         super(Selector,self).__init__(
             default=default, instantiate=instantiate, **params)
         # Required as Parameter sets allow_None=True if default is None
-        self.allow_None = allow_None
-        if default is not None and self.check_on_set is True:
-            self._validate(default)
+        if allow_None is Undefined:
+            self.allow_None = self._slot_defaults['allow_None']
+        if self.default is not None and self.check_on_set is True:
+            self._validate(self.default)
```

Comment:
I don't see why check_on_set would be tested here rather than in self._validate.